service/src/test/java/bio/terra/workspace/service/iam/ControlledResourceSamPolicyBuilderTest.java

```diff
@@ -0,0 +1,134 @@
+package bio.terra.workspace.service.iam;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.contains;
+import static org.junit.jupiter.api.Assertions.*;
+
+import bio.terra.workspace.common.BaseSpringBootUnitTest;
+import bio.terra.workspace.common.exception.InternalLogicException;
+import bio.terra.workspace.common.utils.GcpUtils;
+import bio.terra.workspace.service.iam.model.ControlledResourceIamRole;
+import bio.terra.workspace.service.resource.controlled.model.ControlledResourceCategory;
+import bio.terra.workspace.service.workspace.model.WsmApplication;
+import bio.terra.workspace.service.workspace.model.WsmWorkspaceApplication;
+import org.broadinstitute.dsde.workbench.client.sam.model.CreateResourceRequestV2;
+import org.junit.jupiter.api.Test;
+
+class ControlledResourceSamPolicyBuilderTest extends BaseSpringBootUnitTest {
+
+  @Test
+  void addPolicies_userPrivate() {
+    var policyBuilder =
+        new ControlledResourceSamPolicyBuilder(
+            ControlledResourceIamRole.EDITOR,
+            "fake@example.com",
+            ControlledResourceCategory.USER_PRIVATE,
+            null);
+    var request = new CreateResourceRequestV2();
+
+    policyBuilder.addPolicies(request);
+
+    assertThat(
+        request.getPolicies().get(ControlledResourceIamRole.OWNER.toSamRole()).getMemberEmails(),
+        contains(GcpUtils.getWsmSaEmail()));
+    assertThat(
+        request.getPolicies().get(ControlledResourceIamRole.EDITOR.toSamRole()).getMemberEmails(),
+        contains("fake@example.com"));
+  }
+
+  @Test
+  void addPolicies_userPrivateWihtoutPrivateEmailFails() {
+    var policyBuilder =
+        new ControlledResourceSamPolicyBuilder(
+            ControlledResourceIamRole.DELETER, null, ControlledResourceCategory.USER_PRIVATE, null);
+    var request = new CreateResourceRequestV2();
+
+    assertThrows(InternalLogicException.class, () -> policyBuilder.addPolicies(request));
+  }
+
+  @Test
+  void addPolicies_applicationShared() {
+    var app =
+        new WsmWorkspaceApplication()
+            .application(new WsmApplication().serviceAccount("fake-svc@example.com"));
+    var policyBuilder =
+        new ControlledResourceSamPolicyBuilder(
+            ControlledResourceIamRole.READER,
```

Comment:
What does it mean to specify a READER role but with no user?

Reply:
Good eye; this is a shared resource that is accessible to all user's, so the private IAM role here is actually ignored. This value can be null (or really anything) which points to this being an awkward interface. I'll add some parameter checking to ensure an exception is thrown when a private IAM role is passed but the resource is shared.

Reply:
Update: I've added a logger.warn() to emit on this situation. I had originally put in a stricter check that threw an exception but it caused some worrisome connected test failures that make me worry that there are callsites that are not using this API properly. Rather than refactor the world, I am going to leave that logic as is with the warning, we can reevaluate downstream if it becomes more problematic.